test/test_models.py

```diff
@@ -0,0 +1,369 @@
+        self.assertEqual(type(outputs.data), torch.FloatTensor)
+
+    def test_21_encoder_forward(self):
+        self.encoder_forward(self.opt)
+
+    def test_22_encoder_forward_mean(self):
+        opt = copy.deepcopy(self.opt)
+        opt.encoder_layer = 'mean'
+        self.encoder_forward(opt)
+
+    # def test_23_encoder_forward_transformer(self):
+    #     opt = copy.deepcopy(self.opt)
+    #     opt.encoder_layer = 'transformer'
+    #     opt.word_vec_size = 64
+    #     opt.rnn_size = 64
+    #     self.encoder_forward(opt)
+
+    def test_30_decoder_init(self):
+        vocab = self.get_vocab()
+        onmt.Models.Decoder(self.opt, vocab)
+
+    def test_40_nmtmodel_init(self):
+        vocab = self.get_vocab()
+        enc = onmt.Models.Encoder(self.opt, vocab)
+        dec = onmt.Models.Decoder(self.opt, vocab)
+        onmt.Models.NMTModel(enc, dec)
+
+    def ntmmodel_forward(self, opt, sourceL=3, bsize=1):
+        """
+        Creates a ntmmodel with a custom opt function.
+        Forwards a testbatch anc checks output size.
+
+        Args:
+            opt: Namespace with options
+            sourceL: length of input sequence
+            bsize: batchsize
+        """
+        vocab = self.get_vocab()
+        enc = onmt.Models.Encoder(opt, vocab)
+        dec = onmt.Models.Decoder(opt, vocab)
+        model = onmt.Models.NMTModel(enc, dec)
+
+        test_src, test_tgt, test_length = self.get_batch(sourceL=sourceL,
+                                                         bsize=bsize)
+        outputs, attn, _ = model(test_src,
+                                 test_tgt,
+                                 test_length)
+        outputsize = torch.zeros(sourceL-1, bsize, opt.rnn_size)
+        # Make sure that output has the correct size and type
+        self.assertEqual(outputs.size(), outputsize.size())
+        self.assertEqual(type(outputs), torch.autograd.Variable)
+        self.assertEqual(type(outputs.data), torch.FloatTensor)
+
+    def test_41_nmtmodel_forward(self):
+        """
+        Test to check whether the model forward yields the correct size
+        """
+        self.ntmmodel_forward(self.opt)
+
+    def test_42_nmtmodel_forward_mlp(self):
+        """
+        Test to check whether the model forward yields the correct size
+        with mlp attention
+        """
+        opt = copy.deepcopy(self.opt)
+        opt.attention_type = 'mlp'
+        self.ntmmodel_forward(opt)
+
+    def test_43_nmtmodel_forward_brnn(self):
+        """
+        Test to check whether the model forward yields the correct size
+        with bidirectional encoder
+        """
+        opt = copy.deepcopy(self.opt)
+        opt.brnn = True
+        self.ntmmodel_forward(opt)
+
+    def test_43_1_nmtmodel_forward_brnn_sum(self):
+        """
+        Test to check whether the model forward yields the correct size
+        with bidirectional encoder and merging by sum
+        """
+        opt = copy.deepcopy(self.opt)
```

Comment:
I would rather these not all be individual tests, but more declarative. I.e. just make a list of dictionaries of options and then call ntmmodel_forward on all of them